src/prelude/internals.mo

```diff
@@ -430,15 +443,18 @@ func @install_actor_helper(
         (#reinstall, (prim "cast" : (actor {}) -> Principal) actor1)
       };
       case (#upgrade actor2) {
-        (#upgrade, (prim "cast" : (actor {}) -> Principal) actor2)
+        let upgradeOptions = {
+          wasm_memory_persistence = ?(#Keep);
```

Comment:
This could get tricky in the branch that combines both flavours. It might be better to add another variant, #upgrade_with_persistence (name TBD), of the user facing #upgrade  variant that gives explicit control of the persistence mode, and have the old one retain its original meaning (#Discard).

Reply:
Good point, thanks for noticing. Adding explicit control would be most flexible (especially also for testing specific migration paths).

Reply:
I extended the upgrade functionality accordingly, by adding `#upgrade_with_persistence`. The default persistence mode of `#upgrade` (still supported for backwards compatibility) is selected by the Motoko version (`null` for classical persistence and `#Keep` for enhanced orthogonal persistence). With that, we get an error if the wrong option would be applied to a canister in this auto-selection mode, e.g. 
* When an a canister with classical persistence tries to auto-upgrade (using `#upgrade) a new enhanced orthogonally persistent canister: The upgrade option is then `null` and the IC detects such and rejects the upgrade.
* When a canister with enhanced orthogonal persistence tries to auto-upgrade (using `#upgrade`) a classical canister: The option `#Keep` for memory persistence is not supported for a canister with classical persistence. This is also rejected by the IC.

Reply:
> When an a canister with classical persistence tries to auto-upgrade (using #upgrade) a new enhanced orthogonally persistent canister: The upgrade option is then null` and the IC detects such and rejects the upgrade.
That's not implemented her (it's always `?#Keep`), so I guess we will need to fix this in the final merge with classical-persistence. Maybe by adding a prim that compile-{classical/enhanced}
.ml implements accordingly.

Reply:
Yes, absolutely. In the integrated PR, I use:

```
let wasm_memory_persistence = if enhanced_orthogonal_persistence { 
    ?(#Keep) 
  } else { 
     null 
  };
```